From 3bb2f037ebaceaf51be963bb5b69b49153b787bb Mon Sep 17 00:00:00 2001
From: Anil Kumar Reddy <areddy3@marvell.com>
Date: Tue, 30 Mar 2021 12:18:25 +0530
Subject: [PATCH 1386/1921] hwrng: cn10k: Add random number generator health
 check

This patch add support to check random number generator
health status.

Change-Id: I41f43e0791568b85a2fa7880725faa58f4897aa2
Signed-off-by: Anil Kumar Reddy <areddy3@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/48915
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/char/hw_random/cn10k-rng.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/drivers/char/hw_random/cn10k-rng.c b/drivers/char/hw_random/cn10k-rng.c
index f886bae36cbf..10e2b0ec2845 100644
--- a/drivers/char/hw_random/cn10k-rng.c
+++ b/drivers/char/hw_random/cn10k-rng.c
@@ -24,13 +24,35 @@
 struct cn10k_rng {
 	void __iomem *reg_base;
 	struct hwrng ops;
+	struct pci_dev *pdev;
 };
 
+static int check_rng_health(struct cn10k_rng *rng)
+{
+	u64 status;
+
+	/* Skip checking health */
+	if (!rng->reg_base)
+		return 0;
+
+	status = readq(rng->reg_base + RNM_PF_EBG_HEALTH);
+	if (status & BIT_ULL(20)) {
+		dev_err(&rng->pdev->dev, "HWRNG: Continuous health test failed\n");
+		return -EIO;
+	}
+	return 0;
+}
+
 static int cn10k_rng_read(struct hwrng *hwrng, void *data,
 			  size_t max, bool wait)
 {
 	struct cn10k_rng *rng = (struct cn10k_rng *)hwrng->priv;
 	unsigned int size = max;
+	int err = 0;
+
+	err = check_rng_health(rng);
+	if (err)
+		return err;
 
 	while (size >= 8) {
 		*((u64 *)data) = readq(rng->reg_base + RNM_PF_RANDOM);
@@ -54,6 +76,7 @@ static int cn10k_rng_probe(struct pci_dev *pdev, const struct pci_device_id *id)
 	if (!rng)
 		return -ENOMEM;
 
+	rng->pdev = pdev;
 	pci_set_drvdata(pdev, rng);
 
 	rng->reg_base = pcim_iomap(pdev, 0, 0);
-- 
2.31.1

